### PR TITLE
Shopping Cart service

### PR DESCRIPTION
**Pull request by OlgaStarikova**

### CODE REVIEW

File: src/main/java/com/example/onlinebookstore/model/CartItem.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+import lombok.Setter;
+import org.hibernate.annotations.SQLDelete;
+import org.hibernate.annotations.SQLRestriction;
+
+@Entity
+@Getter
+@Setter
+@SQLDelete(sql = "UPDATE cartitems SET is_deleted = true WHERE id = ?")
+@SQLRestriction("is_deleted = false")
+@Table(name = "cartitems")
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/model/CartItem.java:21]:
```suggestion
@Table(name = "cart_items")
```

File: src/main/java/com/example/onlinebookstore/model/CartItem.java [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+public class CartItem {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @ManyToOne
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/model/CartItem.java:26]:
```suggestion
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
```

File: src/main/java/com/example/onlinebookstore/model/CartItem.java [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    @JoinColumn(nullable = false,name = "cart_id")
+    private ShoppingCart shoppingCart;
+    @ManyToOne
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/model/CartItem.java:29]:
```suggestion
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
```

File: src/main/java/com/example/onlinebookstore/model/ShoppingCart.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+import lombok.Setter;
+import org.hibernate.annotations.SQLDelete;
+import org.hibernate.annotations.SQLRestriction;
+
+@Entity
+@Getter
+@Setter
+@SQLDelete(sql = "UPDATE shoppingcarts SET is_deleted = true WHERE id = ?")
+@SQLRestriction("is_deleted = false")
+@Table(name = "shoppingcarts")
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/model/ShoppingCart.java:23]:
```suggestion
@Table(name = "shopping_carts")
```

File: src/main/java/com/example/onlinebookstore/model/ShoppingCart.java [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+public class ShoppingCart {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @OneToOne
+    @JoinColumn(nullable = false,name = "user_id")
+    private User user;
+    @OneToMany(mappedBy = "shoppingCart")
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/model/ShoppingCart.java:31]:
```suggestion
    @OneToMany(mappedBy = "shoppingCart", cascade = CascadeType.ALL, orphanRemoval = true)
```

File: src/main/java/com/example/onlinebookstore/model/ShoppingCart.java [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+@Getter
+@Setter
+@SQLDelete(sql = "UPDATE shoppingcarts SET is_deleted = true WHERE id = ?")
+@SQLRestriction("is_deleted = false")
+@Table(name = "shoppingcarts")
+public class ShoppingCart {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @OneToOne
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/model/ShoppingCart.java:28]:
```suggestion
    @OneToOne(fetch = FetchType.LAZY)
```

File: src/main/java/com/example/onlinebookstore/service/CartService.java [hunk 7]
```diff
@@ -0,0 +1,20 @@
+package com.example.onlinebookstore.service;
+
+import com.example.onlinebookstore.dto.CartItemDto;
+import com.example.onlinebookstore.dto.CreateCartItemRequestDto;
+import com.example.onlinebookstore.dto.ShoppingCartDto;
+import com.example.onlinebookstore.dto.UpdateCartItemRequestDto;
+import com.example.onlinebookstore.model.ShoppingCart;
+
+public interface CartService {
+
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/service/CartService.java:10]:
```suggestion
```

File: src/main/java/com/example/onlinebookstore/service/impl/CartServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+@Service
+public class CartServiceImpl implements CartService {
+    private final BookRepository bookRepository;
+    private final CartItemRepository cartItemRepository;
+    private final ShoppingCartRepository shoppingCartRepository;
+    private final ShoppingCartMapper cartMapper;
+    private final UserRepository userRepository;
+
+    @Override
+    public CartItemDto save(String email, CreateCartItemRequestDto requestDto) {
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/service/impl/CartServiceImpl.java:30]:
```suggestion
    public ShoppingCartDto save(String email, CreateCartItemRequestDto requestDto) {
```

File: src/main/java/com/example/onlinebookstore/service/impl/CartServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+        CartItem cartItem = cartMapper.toCartItemModel(requestDto);
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/service/impl/CartServiceImpl.java:31]: let's make your code more readable
Create Book book = repo.find
ShoppingCart cart = repo.find
cart.getCartItems() -> add new cartItem, check if there is already bookId exists then increase quantity
if no -> add new cartItem

File: src/main/java/com/example/onlinebookstore/service/impl/UserServiceImpl.java [hunk 10]
```diff
@@ -32,6 +35,10 @@ public UserResponseDto register(UserRegistrationRequestDto requestDto) {
         User user = userMapper.toModel(requestDto);
         user.setPassword(passwordEncoder.encode(requestDto.password()));
         user.setRoles(Set.of(roleRepository.getByRole(Role.RoleName.USER)));
-        return userMapper.toDto(userRepository.save(user));
+        User savedUser = userRepository.save(user);
+        ShoppingCart cart = new ShoppingCart();
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/service/impl/UserServiceImpl.java:39]: It's better to create method createShoppingCart(User user) in ShopCartService and call it here

File: src/main/java/com/example/onlinebookstore/service/impl/CartServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+        cartItem.setShoppingCart(findShoppingCartByUserEmail(email));
+        cartItem.setBook(bookRepository.findById(requestDto.bookId())
+                .orElseThrow(() -> new EntityNotFoundException(
+                        "No such book with id: " + requestDto.bookId())));
+        return cartMapper.toCartItemDto(cartItemRepository.save(cartItem));
+    }
+
+    @Override
+    public CartItemDto update(Long id, UpdateCartItemRequestDto requestDto) {
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/service/impl/CartServiceImpl.java:40]:
```suggestion
    public ShoppingCartDto update(Long id, UpdateCartItemRequestDto requestDto) {
```

File: src/main/resources/db/changelog/changes/10-create-shoppingcart-table.yml [hunk 12]
```diff
@@ -0,0 +1,29 @@
+databaseChangeLog:
+  - changeSet:
+      id: create-shoppingcart-table
+      author: Starykova Olga
+      changes:
+        - createTable:
+            tableName: shoppingcarts
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/changes/10-create-shoppingcart-table.yml:7]: Check comments in Entity

File: src/main/resources/db/changelog/changes/11-create-cartitem-table.yml [hunk 13]
```diff
@@ -0,0 +1,40 @@
+databaseChangeLog:
+  - changeSet:
+      id: create-cartitems-table
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/changes/11-create-cartitem-table.yml:3]:
```suggestion
      id: create-cart-items-table
```

File: src/main/resources/db/changelog/changes/11-create-cartitem-table.yml [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+                  type: bigint
+                  autoIncrement: true
+                  constraints:
+                    primaryKey: true
+                    nullable: false
+              - column:
+                  name: cart_id
+                  type: bigint
+                  constraints:
+                    foreignKeyName: fk_cartitems_shoppingcarts
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/changes/11-create-cartitem-table.yml:20]:
```suggestion
                    foreignKeyName: fk_cart_items_shoppingcarts
```
Check all places

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
Let's fix some comments

File: src/main/java/com/example/onlinebookstore/dto/CreateCartItemRequestDto.java [hunk 15]
```diff
@@ -0,0 +1,12 @@
+package com.example.onlinebookstore.dto;
+
+import jakarta.validation.constraints.Min;
+import jakarta.validation.constraints.NotNull;
+
+public record CreateCartItemRequestDto(
+        @NotNull
+        Long bookId,
+        @Min(1)
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/dto/CreateCartItemRequestDto.java:9]:
```suggestion
        @Positive
```

File: src/main/java/com/example/onlinebookstore/dto/UpdateCartItemRequestDto.java [hunk 16]
```diff
@@ -0,0 +1,9 @@
+package com.example.onlinebookstore.dto;
+
+import jakarta.validation.constraints.Min;
+
+public record UpdateCartItemRequestDto(
+        @Min(1)
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/dto/UpdateCartItemRequestDto.java:6]:
```suggestion
        @Positive
```

File: src/main/java/com/example/onlinebookstore/mapper/ShoppingCartMapper.java [hunk 17]
```diff
@@ -0,0 +1,28 @@
...
+
+@Mapper(config = MapperConfig.class)
+public interface ShoppingCartMapper {
+    @Mapping(source = "user.id", target = "userId")
+    @Mapping(source = "cartItems", target = "cartItems")
+    ShoppingCartDto toShoppingCartDto(ShoppingCart shoppingCart);
+
+    @Mapping(source = "book.id", target = "bookId")
+    @Mapping(source = "book.title", target = "bookTitle")
+    CartItemDto toCartItemDto(CartItem cartItem);
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/mapper/ShoppingCartMapper.java:22]: It's better to create CartItemMapper to separate logic from ShoppingCartMapper

File: src/main/java/com/example/onlinebookstore/service/impl/CartServiceImpl.java [hunk 18]
```diff
@@ -0,0 +1,86 @@
...
+            cartItemRepository.save(cartItem);
+        }
+        shoppingCart.setCartItems(cartItemRepository
+                .findCartItemsByShoppingCart_Id(shoppingCart.getId()));
+        return cartMapper.toShoppingCartDto(shoppingCart);
+    }
+
+    @Override
+    public ShoppingCartDto update(String email, Long id, UpdateCartItemRequestDto requestDto) {
+        CartItem cartItem = cartItemRepository.findById(id)
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/service/impl/CartServiceImpl.java:56]:
```suggestion
        ShoppingCart cart = repository.findByUserId(userId);
        CartItem cartItem = itemRepository.findByIdAndShoppingCartId(cartItemId, cart.getId())
                .map(item -> {
                    item.setQuantity(itemDto.quantity());
                    return item;
                }).orElseThrow(() -> new EntityNotFoundException(
                        String.format("No cart item with id: %d for user: %d", cartItemId, userId)
                ));
        itemRepository.save(cartItem);
        return cartMapper.toDto(cart);
```

File: src/main/java/com/example/onlinebookstore/service/impl/CartServiceImpl.java [hunk 19]
```diff
@@ -0,0 +1,86 @@
...
+    public void deleteById(Long id) {
+        cartItemRepository.deleteById(id);
+    }
+
+    public ShoppingCartDto getShoppingCartByUserEmail(String email) {
+        return cartMapper.toShoppingCartDto(findShoppingCartByUserEmail(email));
+    }
+
+    @Override
+    public ShoppingCart createShoppingCart(ShoppingCart shoppingCart) {
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/service/impl/CartServiceImpl.java:75]:
```suggestion
    public void createShoppingCart(User user) {
        ShoppingCart shoppingCart = new ShoppingCart();
        shoppingCart.setUser(user);
        repository.save(shoppingCart);
    }
```

File: src/main/java/com/example/onlinebookstore/service/impl/UserServiceImpl.java (hunk 10 above)
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/service/impl/UserServiceImpl.java:39]:
```suggestion
        cartService.createShoppingCart(user);
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
leave comments

File: src/main/java/com/example/onlinebookstore/controller/ShoppingCartController.java [hunk 20]
```diff
@@ -0,0 +1,64 @@
...
+
+@RestController
+@Tag(name = "ShoppingCart management", description = "Endpoints for managing shopping carts")
+@RequestMapping("/cart")
+@RequiredArgsConstructor
+public class ShoppingCartController {
+    private final CartService cartService;
+
+    @GetMapping
+    public ShoppingCartDto getShoppingCart() {
```
Comment (okuzan) [inline src/main/java/com/example/onlinebookstore/controller/ShoppingCartController.java:32]: add @ Operation here too

File: src/main/java/com/example/onlinebookstore/dto/CreateCartItemRequestDto.java [hunk 21]
```diff
@@ -0,0 +1,12 @@
+package com.example.onlinebookstore.dto;
+
+import jakarta.validation.constraints.NotNull;
+import jakarta.validation.constraints.Positive;
+
+public record CreateCartItemRequestDto(
+        @NotNull
+        Long bookId,
```
Comment (okuzan) [inline src/main/java/com/example/onlinebookstore/dto/CreateCartItemRequestDto.java:8]: you can use @ Positive on long ids in request DTOs too

File: src/main/java/com/example/onlinebookstore/model/CartItem.java [hunk 22]
```diff
@@ -0,0 +1,37 @@
...
+import lombok.Getter;
+import lombok.Setter;
+import org.hibernate.annotations.SQLDelete;
+import org.hibernate.annotations.SQLRestriction;
+
+@Entity
+@Getter
+@Setter
+@SQLDelete(sql = "UPDATE cart_items SET is_deleted = true WHERE id = ?")
+@SQLRestriction("is_deleted = false")
```
Comment (okuzan) [inline src/main/java/com/example/onlinebookstore/model/CartItem.java:21]:
```suggestion
@SQLDelete(sql = "UPDATE cart_items SET is_deleted = TRUE WHERE id = ?")
@SQLRestriction("is_deleted = FALSE")
```
they are SQL keywords too

File: src/main/java/com/example/onlinebookstore/service/impl/CartServiceImpl.java [hunk 23]
```diff
@@ -0,0 +1,96 @@
...
+    }
+
+    @Override
+    public void deleteById(Long id) {
+        cartItemRepository.deleteById(id);
+    }
+
+    public ShoppingCartDto getShoppingCartByUserEmail(String email) {
+        return cartMapper.toShoppingCartDto(findShoppingCartByUserEmail(email));
+    }
```
Comment (okuzan) [inline src/main/java/com/example/onlinebookstore/service/impl/CartServiceImpl.java:80]: make this private
put it after all public methods

File: src/main/resources/db/changelog/db.changelog-master.yaml [hunk 24]
```diff
@@ -16,4 +16,8 @@ databaseChangeLog:
   - include:
       file: db/changelog/changes/08-create-category-table.yml
   - include:
-      file: db/changelog/changes/09-create-book-category-table.yml
\ No newline at end of file
+      file: db/changelog/changes/09-create-book-category-table.yml
+  - include:
+      file: db/changelog/changes/10-create-shoppingcart-table.yml
+  - include:
+      file: db/changelog/changes/11-create-cartitem-table.yml
```
Comment (okuzan) [inline src/main/resources/db/changelog/db.changelog-master.yaml:23]: add empty line

File: src/main/resources/db/changelog/changes/11-create-cartitem-table.yml [hunk 25]
```diff
@@ -0,0 +1,40 @@
...
+                    nullable: false
+              - column:
+                  name: quantity
+                  type: int
+              - column:
+                  name: is_deleted
+                  type: boolean
+                  defaultValueBoolean: false
+                  constraints:
+                    nullable: false
```
Comment (okuzan) [inline src/main/resources/db/changelog/changes/11-create-cartitem-table.yml:40]: add empty line

**Review by okuzan — CHANGES_REQUESTED** [review]

**Review by Elena-Bruyako — APPROVED** [review]
Well done!

**Pull request merged**